src/main/kotlin/com/tradeshift/blayze/features/Multinomial.kt

```diff
@@ -13,76 +12,124 @@ import kotlin.math.pow
+        for (vec in features.values) {
+            for ((idx, count) in vec) {
+                res[idx] += count
+            }
+        }
+        res
+    }
 
     override fun batchUpdate(updates: List<Pair<Outcome, Counter<String>>>): Multinomial {
-        val updatedTable = countTable.toMutableTable()
+        val featuresCopy = features.toMutableMap()
+        val outcomesCopy = outcomes.toMutableMap()
 
-        val knownFeatures = updatedTable.columnKeySet.toMutableSet()
-        for ((outcome, counts) in updates) {
-            for ((feature, value) in counts) {
-                if (knownFeatures.contains(feature) || Math.random() < (1.0 - (1.0 - includeFeatureProbability).pow(value))) {
-                    knownFeatures.add(feature)
-                    val count = updatedTable[outcome, feature] ?: 0
-                    updatedTable.put(outcome, feature, count + value)
-                }
-            }
+        fun getOrCreateIndex(key: Outcome) = outcomesCopy.getOrPut(key, { outcomesCopy.size })
+
+        val formattedUpdates = invertUpdates(sampleUpdates(updates.asSequence()))
+        for ((feature, counter) in formattedUpdates) {
+            val indexToUpdate = counter.mapKeys { getOrCreateIndex(it.key)  }
+            val vec = SparseIntVector.fromMap(indexToUpdate)
+            featuresCopy[feature] = featuresCopy[feature]?.add(vec) ?: vec
         }
-        return Multinomial(includeFeatureProbability, pseudoCount, updatedTable.toTable())
+        return Multinomial(includeFeatureProbability, pseudoCount, outcomesCopy, featuresCopy)
     }
 
     override fun logProbability(outcomes: Set<Outcome>, value: Counter<String>): Map<Outcome, Double> {
-        val knownFeatures: Set<String> = value.keys.intersect(countTable.columnKeySet)
-        val numOfFeatures = countTable.columnKeySet.size
-
-        val results = mutableMapOf<Outcome, Double>()
-        for (outcome in outcomes) {
-            var logProb = 0.0
-            for (feature in knownFeatures) {
-                val count = ((countTable[outcome, feature] ?: 0) + pseudoCount)
-                logProb += value[feature] * (ln(count) - ln(outcomeCounter[outcome] + (numOfFeatures * pseudoCount)))
+        // sum all log probabilities for non-zero feature-outcome combinations
+        val nonZeroLogProbs = DoubleArray(this.outcomes.size)
+        val nonZeroCounts = IntArray(this.outcomes.size)
+        var nFeatures = 0
+        for ((feature, count) in value) {
+            val vector = features[feature]
+            if (vector != null) {
+                nFeatures += count
+                for ((idx, v) in vector) {
+                    nonZeroLogProbs[idx] += count * logProbability(v, outcomeCounts[idx])
+                    nonZeroCounts[idx] += count
+                }
             }
-            results[outcome] = logProb
         }
-        return results
+
+        // then compute log probabilities for zero valued feature-outcome combinations
```

Comment:
It's pretty clever, but also kind of hard to understand the way nonZero and zeroCounts are split. 
Can't you declare logProps directly, and then inside the loop:
 * Add the nonZero log props
 * Add the zero log props

Reply:
Nevermind. I don't think I fully understand your code and why it works.

Reply:
The main prediction speedup comes from only counting the zero values once per outcome class. Moving it inside the loop would mean counting them per outcome class per input feature which is a equivalent to counting up dense vectors.